src/go/rpk/pkg/config/manager.go

```diff
@@ -323,7 +352,14 @@ func (m *manager) Set(key, value, format, path string) error {
 	}
 	newV := viper.New()
 	newV.Set(key, newConfValue)
-	m.v.MergeConfigMap(newV.AllSettings())
+	src = newV.AllSettings()
+	dst = m.v.AllSettings()
+	merged, err = mergeMaps(dst, src)
+	if err != nil {
+		return err
+	}
+	m.v = InitViper(m.fs)
```

Comment:
why not `m.MergeConfigMap( foo )` does a `m.Reset() + m.v.MergeConfigMap` - if you raise it up a level it will always be correct.

Reply:
Yup, that's the next step. This is still a draft hehe :D

I'm also questioning viper's importance here at this point. The main thing was using it to bind the flags, but then we found that bug, and now it's kind of a wrapper for a config `map` (with `Write` & `Read` goodies). Probably not for this PR, but defs something I'll be looking into.

Reply:
Yup, that's the next step. This is still a draft hehe :D

I'm also questioning viper's importance here at this point. The main thing was using it to bind the flags, but then we found that bug, and now it's kind of a wrapper for a config `map` (with `Write` & `Read` goodies). Probably not for this PR, but defs something I'll be looking into.

Reply:
I had forgotten to publish that comment last Thursday 🤦🏽‍♂️  ^

I made changes that I think incorporate what you were suggesting. Please let me know!